relay-server/src/actors/spooler/mod.rs

```diff
@@ -525,33 +745,25 @@ impl BufferService {
             sender,
         } = message;
 
-        for key in &keys {
-            for value in self.buffer.remove(key).unwrap_or_default() {
-                self.used_memory -= value.estimated_size();
-                sender.send(value).ok();
+        match self.state {
+            BufferState::Memory { ref mut ram }
+            | BufferState::MemoryFileStandby { ref mut ram, .. } => {
+                ram.dequeue(&keys, sender);
+            }
+            BufferState::File { ref disk } => {
+                disk.dequeue(project_key, &mut keys, sender, &self.services)
+                    .await;
+                self.state.update(&self.config).await?;
+            }
+            BufferState::MemoryFileRead {
+                ref mut ram,
+                ref disk,
+            } => {
+                ram.dequeue(&keys, sender.clone());
+                disk.dequeue(project_key, &mut keys, sender, &self.services)
+                    .await;
+                self.state.update(&self.config).await?;
```

Comment:
nit: Same as above, could call `self.state.update()` after the `match` statement.